libflux/go/libflux/parser_test.go

```diff
@@ -118,29 +118,32 @@ c = 3`)
 			name:    "packages clauses don't match",
 			outPkg:  barPkg,
 			inPkg:   fooPkg,
-			wantErr: errors.New("failed to merge packages: file's package clause: foo does not match package output package clause: bar"),
+			wantErr: errors.New(`failed to merge packages: file's package clause: "foo" does not match package output package clause: "bar"`),
 		},
 		{
 			name:    "input package has no package clause",
 			outPkg:  barPkg,
 			inPkg:   noClausePkg,
-			wantErr: errors.New("failed to merge packages: current file does not have a package clause"),
+			wantErr: errors.New(`failed to merge packages: current file does not have a package clause, but output package has package clause "bar"`),
 		},
 		{
 			name:    "output package has no package clause",
 			outPkg:  noClausePkg,
 			inPkg:   fooPkg,
-			wantErr: errors.New("failed to merge packages: output package does not have a package clause"),
+			wantErr: errors.New(`failed to merge packages: output package does not have a package clause, but current file has package clause "foo"`),
 		},
 	}
 
 	for _, testCase := range testCases {
-		gotErr := libflux.MergePackages(testCase.outPkg, testCase.inPkg)
-		if gotErr == nil {
-			t.Fatal("\nGot no error, expected:", testCase.wantErr, "\ngot: \n", gotErr)
-		}
-		if diff := cmp.Diff(testCase.wantErr.Error(), gotErr.Error()); diff != "" {
-			t.Fatalf("unexpected error: -want/+got: %v", diff)
-		}
+		testCase := testCase
```

Comment:
I'm curious about this assignment here, is there a reason to create this variable?

Reply:
It's a technique for creating a local variable that gets captured. When you use a for loop, the memory it assigns to that variable is reused for each iteration as a way of making the for loop faster. When you use a function inside that has any parallelism, this can result in having each test use only the last value in the array instead of what would normally be expected.

To my knowledge, this is only needed when `t.Parallel()` is used since that causes the `Run` to not be executed immediately. But there's also no harm to doing it whenever you use subtests.

Reply:
The reason is that `go test` will run some test cases in parallel, if `t.Parallel()` is called.  I guess we are not doing that here.  But if we did, then a goroutine that reads `testCase` from inside the testing `func(t *testing.T) {...}` might see the value of `testCase` change unexpectedly, and lead to non-deterministic behavior.  So, by doing `testCase := testCase`, we create a copy that can be used in the test function safely.

I have just gotten into the habit of doing it, since I've seen it elsewhere.

Reply:
I figured it had to do with copying the local variable, but was curious as to why. Thanks for the explainers.